pkg/kubectl/cmd/apply.go

```diff
@@ -215,3 +202,77 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
 
 	return nil
 }
+
+type patcher struct {
+	encoder runtime.Encoder
+	decoder runtime.Decoder
+
+	mapping *meta.RESTMapping
+	helper  *resource.Helper
+
+	backOff *flowcontrol.Backoff
+}
+
+func NewPatcher(encoder runtime.Encoder, decoder runtime.Decoder, mapping *meta.RESTMapping, helper *resource.Helper) *patcher {
+	return &patcher{
+		encoder: encoder,
+		decoder: decoder,
+		mapping: mapping,
+		helper:  helper,
+		backOff: flowcontrol.NewBackOff(backOffPeriod, maxApplyBackOff),
+	}
+}
+
+func (p *patcher) patchSimple(obj runtime.Object, modified []byte, source, namespace, name string) ([]byte, error) {
+	// Serialize the current configuration of the object from the server.
+	current, err := runtime.Encode(p.encoder, obj)
+	if err != nil {
+		return nil, cmdutil.AddSourceToErr(fmt.Sprintf("serializing current configuration from:\n%v\nfor:", obj), source, err)
+	}
+
+	// Retrieve the original configuration of the object from the annotation.
+	original, err := kubectl.GetOriginalConfiguration(p.mapping, obj)
+	if err != nil {
+		return nil, cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", obj), source, err)
+	}
+
+	// Create the versioned struct from the original from the server for
+	// strategic patch.
+	// TODO: Move all structs in apply to use raw data. Can be done once
+	// builder has a RawResult method which delivers raw data instead of
+	// internal objects.
+	versionedObject, _, err := p.decoder.Decode(current, nil, nil)
+	if err != nil {
+		return nil, cmdutil.AddSourceToErr(fmt.Sprintf("converting encoded server-side object back to versioned struct:\n%v\nfor:", obj), source, err)
+	}
+
+	// Compute a three way strategic merge patch to send to server.
+	patch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, versionedObject, true)
+	if err != nil {
+		format := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfor:"
+		return nil, cmdutil.AddSourceToErr(fmt.Sprintf(format, original, modified, current), source, err)
+	}
+
+	_, err = p.helper.Patch(namespace, name, api.StrategicMergePatchType, patch)
+	return patch, err
+}
+
+func (p *patcher) patch(current runtime.Object, modified []byte, source, namespace, name string) ([]byte, error) {
+	var getErr error
+	patchBytes, err := p.patchSimple(current, modified, source, namespace, name)
+	for i := 0; i < maxPatchRetry && errors.IsConflict(err); i++ {
+		if p.backOff.IsInBackOffSinceUpdate(backOffKey, p.backOff.Clock.Now()) {
```

Comment:
wow, this is a cool utility.  I think it may be overkill for what we are trying to do here, since it seems primarily for synchornizing ratelimits across concurrent routines.  At a minimum, we should sleep the backoff duration (sleep p.backOff.Get(backOffKey)).  While I really appreciate you showing me this utility, if we don't need the concurrency protection, I think I'd prefer to go with something simpler such as:

```go
Patcher {
...
backoff clockwork.Clock // = clockwork.NewRealClock()
}

...

p.backoff.Sleep(p.duration)

```